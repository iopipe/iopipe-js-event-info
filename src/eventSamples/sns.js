export default {
  Records: [
    {
      EventVersion: '1.0',
      EventSource: 'aws:sns',
      EventSubscriptionArn:
        'arn:aws:sns:us-west-2:123456789000:ses_messages:26a58451-3392-4ab6-a829-d65c2968421a',
      Sns: {
        MessageId: '483eae4c-4fb0-57e5-a5f9-ff9b08612bef',
        Signature:
          'Uy3tn/qAQg/sXARGk2DRddd31ZtyDE+B1IzRla/KA75BaerApJqN+H59q69z8H+pRx0AyUwOD1K0huBYdDRbAMVOUsMgZgdcNjj0gSfFg8uZvTuKaqTaWj4E0hmzoemHENWeuswuq3l6xoPcAJ9fHd2yFhX+792AV++i/8P4EKv/9I4j8Ejs3OxMRN49gkWefKbv4/avyHOdSaFTnXV0rGLmPb103dtjeY4K05PTKvUlPerN+MdRTvHrjApvqDvP0NEVyYBU4zFZQ6GnFcFnHtTk44c3NH/dVi6Gf9VrX8V1id5VSZICYiIG1iaUZ0b676IhRh8znzjMDWaczOBwkA==',
        Type: 'Notification',
        TopicArn: 'arn:aws:sns:us-west-2:123456789000:ses_messages',
        SignatureVersion: '1',
        Timestamp: '2017-07-05T20:01:21.366Z',
        SigningCertUrl:
          'https://sns.us-west-2.amazonaws.com/SimpleNotificationService-b95095beb82e8f6a046b3aafc7f4149a.pem',
        UnsubscribeUrl:
          'https://sns.us-west-2.amazonaws.com/?Action=Unsubscribe&eifjccgihujihfhrchunfnglreichbrcljrnlvtbeked',
        Subject: 'This is a test subject'
      }
    }
  ]
};
