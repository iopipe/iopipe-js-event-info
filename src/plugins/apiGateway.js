import logFromKeys from '../util/logFromKeys';

const type = 'apiGateway';

function eventType(event) {
  const apiGatewayRequiredKeys = [
    'path',
    'headers',
    'requestContext',
    'resource',
    'httpMethod'
  ];
  if (typeof event === 'object') {
    return apiGatewayRequiredKeys.every(s => s in event) ? type : false;
  }
  return false;
}

const keys = [
  'httpMethod',
  'path',
  'requestContext.accountId',
  'requestContext.httpMethod',
  'requestContext.identity.userAgent',
  'requestContext.requestId',
  'requestContext.resourcePath',
  'requestContext.stage',
  'resource'
];

function plugin(event, log) {
  logFromKeys({ event, type, keys, log });
}

export { eventType, plugin };
