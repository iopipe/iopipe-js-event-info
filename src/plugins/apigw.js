module.exports = function handleApiGwEvent(event, log) {
  const requiredKeys = [
    'path',
    'headers',
    'requestContext',
    'resource',
    'httpMethod'
  ];
  for (let i = 0; i < requiredKeys.length; i++) {
    if (!(requiredKeys[i] in event)) {
      return;
    }
  }
  log('event-apigw-path', event.path);
  log('event-apigw-accountId', event.requestContext.accountId);
  log('event-apigw-stage', event.requestContext.stage);
  log('event-apigw-requestId', event.requestContext.requestId);
  if (event.requestContext.identity) {
    log(
      'event-apigw-identityUserAgent',
      event.requestContext.identity.userAgent
    );
  }
  log('event-apigw-requestResourcePath', event.requestContext.resourcePath);
  log('event-apigw-requestHttpMethod', event.requestContext.httpMethod);
  log('event-apigw-httpMethod', event.httpMethod);
  log('event-apigw-resource', event.resource);
};
