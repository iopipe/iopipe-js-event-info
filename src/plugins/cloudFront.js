import logFromKeys from '../util/logFromKeys';

const type = 'cloudFront';

function eventType(event = {}) {
  const { Records = [] } = event;
  return Records[0] && Records[0].cf ? type : false;
}

const keys = [
  'config.distributionId',
  'request.clientIp',
  'request.method',
  'request.uri',
  'request.headers.host[0].value',
  'request.headers.["user-agent"][0].value'
].map(str => `Records[0].cf.${str}`);

function plugin(event, log) {
  logFromKeys({
    type,
    event,
    keys,
    log
  });
}

export { eventType, plugin };
