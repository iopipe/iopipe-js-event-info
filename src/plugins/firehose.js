import logFromKeys from '../util/logFromKeys';

const type = 'firehose';

function eventType(event = {}) {
  const { records = [] } = event;
  return event.deliveryStreamArn &&
    records[0] &&
    records[0].kinesisRecordMetadata
    ? type
    : false;
}

const keys = ['records.length', 'deliveryStreamArn', 'region'];

function plugin(event, log) {
  logFromKeys({
    type,
    event,
    keys,
    log
  });
}

export { eventType, plugin };
