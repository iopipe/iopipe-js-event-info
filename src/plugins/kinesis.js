import logFromKeys from '../util/logFromKeys';

const type = 'kinesis';

function eventType(event = {}) {
  const { Records = [] } = event;
  const [firstEvent = {}] = Records;
  const { eventVersion, eventSource } = firstEvent;
  return eventVersion === '1.0' && eventSource === 'aws:kinesis' ? type : false;
}

const keys = [
  'Records[0].awsRegion',
  'Records[0].eventSourceARN',
  'Records.length'
];

function plugin(event, log) {
  logFromKeys({
    type,
    event,
    keys,
    log
  });
}

export { eventType, plugin };
