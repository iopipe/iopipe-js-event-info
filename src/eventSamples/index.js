import apiGateway from './apiGateway';
import cloudFront from './cloudFront';
import firehose from './firehose';
import kinesis from './kinesis';
import s3 from './s3';
import scheduled from './scheduled';
import sns from './sns';

export { apiGateway, cloudFront, firehose, kinesis, s3, scheduled, sns };
