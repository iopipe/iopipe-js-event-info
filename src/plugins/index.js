import * as apiGateway from './apiGateway';
import * as cloudFront from './cloudFront';
import * as firehose from './firehose';
import * as kinesis from './kinesis';
import * as s3 from './s3';
import * as scheduled from './scheduled';
import * as sns from './sns';

export { apiGateway, cloudFront, firehose, kinesis, s3, scheduled, sns };
