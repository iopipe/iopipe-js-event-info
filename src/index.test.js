import _ from 'lodash';
import eventInfoPlugin from './index';

import * as miniPlugins from './plugins';
import * as eventSamples from './eventSamples';

class MockInvocation {
  constructor(event) {
    this.logData = {};
    this.context = {
      iopipe: {
        log: (key, value) => {
          this.logData[key] = value;
        }
      }
    };
    this.event = event;
    return this;
  }
}

_.keys(miniPlugins).map(pluginName => {
  return test(`${pluginName} reports snapshot`, () => {
    /*eslint-disable import/namespace*/
    const sampleRecord = eventSamples[pluginName];
    /*eslint-enable import/namespace*/
    const invocationInstance = new MockInvocation(sampleRecord);
    eventInfoPlugin()(invocationInstance).preReport();
    const { logData } = invocationInstance;
    expect(_.isEmpty(logData)).toBe(false);
    expect(logData).toMatchSnapshot();
  });
});
