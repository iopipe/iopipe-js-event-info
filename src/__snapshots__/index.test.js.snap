// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiGateway reports snapshot 1`] = `
Object {
  "@iopipe/event-info.apiGateway.httpMethod": "GET",
  "@iopipe/event-info.apiGateway.path": "/test/hello",
  "@iopipe/event-info.apiGateway.requestContext.accountId": "123456789012",
  "@iopipe/event-info.apiGateway.requestContext.httpMethod": "GET",
  "@iopipe/event-info.apiGateway.requestContext.identity.userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.82 Safari/537.36 OPR/39.0.2256.48",
  "@iopipe/event-info.apiGateway.requestContext.requestId": "41b45ea3-70b5-11e6-b7bd-69b5aaebc7d9",
  "@iopipe/event-info.apiGateway.requestContext.resourcePath": "/{proxy+}",
  "@iopipe/event-info.apiGateway.requestContext.stage": "test",
  "@iopipe/event-info.apiGateway.resource": "/{proxy+}",
  "@iopipe/event-info.eventType": "apiGateway",
}
`;

exports[`cloudFront reports snapshot 1`] = `
Object {
  "@iopipe/event-info.cloudFront.Records[0].cf.config.distributionId": "EDFDVBD6EXAMPLE",
  "@iopipe/event-info.cloudFront.Records[0].cf.request.clientIp": "2001:0db8:85a3:0:0:8a2e:0370:7334",
  "@iopipe/event-info.cloudFront.Records[0].cf.request.headers.[\\"user-agent\\"][0].value": "curl/7.51.0",
  "@iopipe/event-info.cloudFront.Records[0].cf.request.headers.host[0].value": "d111111abcdef8.cloudfront.net",
  "@iopipe/event-info.cloudFront.Records[0].cf.request.method": "GET",
  "@iopipe/event-info.cloudFront.Records[0].cf.request.uri": "/picture.jpg",
  "@iopipe/event-info.eventType": "cloudFront",
}
`;

exports[`firehose reports snapshot 1`] = `
Object {
  "@iopipe/event-info.eventType": "firehose",
  "@iopipe/event-info.firehose.deliveryStreamArn": "aws:lambda:events",
  "@iopipe/event-info.firehose.records.length": 2,
  "@iopipe/event-info.firehose.region": "us-west-2",
}
`;

exports[`kinesis reports snapshot 1`] = `
Object {
  "@iopipe/event-info.eventType": "kinesis",
  "@iopipe/event-info.kinesis.Records.length": 1,
  "@iopipe/event-info.kinesis.Records[0].awsRegion": "us-east-1",
  "@iopipe/event-info.kinesis.Records[0].eventSourceARN": "event-source-arn",
}
`;

exports[`s3 reports snapshot 1`] = `
Object {
  "@iopipe/event-info.eventType": "s3",
  "@iopipe/event-info.s3.Records[0].awsRegion": "us-east-1",
  "@iopipe/event-info.s3.Records[0].eventName": "event-type",
  "@iopipe/event-info.s3.Records[0].eventTime": "1970-01-01T00:00:00.000Z",
  "@iopipe/event-info.s3.Records[0].requestParameters.sourceIPAddress": "ip-address-where-request-came-from",
  "@iopipe/event-info.s3.Records[0].responseElements[\\"x-amz-id-2\\"]": "Amazon S3 host that processed the request",
  "@iopipe/event-info.s3.Records[0].responseElements[\\"x-amz-request-id\\"]": "Amazon S3 generated request ID",
  "@iopipe/event-info.s3.Records[0].s3.bucket.arn": "bucket-ARN",
  "@iopipe/event-info.s3.Records[0].s3.bucket.name": "bucket-name",
  "@iopipe/event-info.s3.Records[0].s3.object.key": "object-key",
  "@iopipe/event-info.s3.Records[0].s3.object.sequencer": "a string representation of a hexadecimal value used to determine event sequence, only used with PUTs and DELETEs",
  "@iopipe/event-info.s3.Records[0].s3.object.size": 10,
  "@iopipe/event-info.s3.Records[0].userIdentity.principalId": "Amazon-customer-ID-of-the-user-who-caused-the-event",
}
`;

exports[`scheduled reports snapshot 1`] = `
Object {
  "@iopipe/event-info.eventType": "scheduled",
  "@iopipe/event-info.scheduled.account": "123456789012",
  "@iopipe/event-info.scheduled.id": "cdc73f9d-aea9-11e3-9d5a-835b769c0d9c",
  "@iopipe/event-info.scheduled.region": "us-east-1",
  "@iopipe/event-info.scheduled.resources[0]": "arn:aws:events:us-east-1:123456789012:rule/my-schedule",
  "@iopipe/event-info.scheduled.time": "1970-01-01T00:00:00Z",
}
`;

exports[`sns reports snapshot 1`] = `
Object {
  "@iopipe/event-info.eventType": "sns",
  "@iopipe/event-info.sns.Records[0].EventSubscriptionArn": "arn:aws:sns:us-west-2:123456789000:ses_messages:26a58451-3392-4ab6-a829-d65c2968421a",
  "@iopipe/event-info.sns.Records[0].Sns.MessageId": "483eae4c-4fb0-57e5-a5f9-ff9b08612bef",
  "@iopipe/event-info.sns.Records[0].Sns.Signature": "Uy3tn/qAQg/sXARGk2DRddd31ZtyDE+B1IzRla/KA75BaerApJqN+H59q69z8H+pRx0AyUwOD1K0huBYdDRbAMVOUsMgZgdcNjj0gSfFg8uZvTuKaqTaWj4E0hmzoemHENWeuswuq3l6xoPcAJ9fHd2yFhX+792AV++i/8P4EKv/9I4j8Ejs3OxMRN49gkWefKbv4/avyHOdSaFTnXV0rGLmPb103dtjeY4K05PTKvUlPerN+MdRTvHrjApvqDvP0NEVyYBU4zFZQ6GnFcFnHtTk44c3NH/dVi6Gf9VrX8V1id5VSZICYiIG1iaUZ0b676IhRh8znzjMDWaczOBwkA==",
  "@iopipe/event-info.sns.Records[0].Sns.SignatureVersion": "1",
  "@iopipe/event-info.sns.Records[0].Sns.SigningCertUrl": "https://sns.us-west-2.amazonaws.com/SimpleNotificationService-b95095beb82e8f6a046b3aafc7f4149a.pem",
  "@iopipe/event-info.sns.Records[0].Sns.Subject": "This is a test subject",
  "@iopipe/event-info.sns.Records[0].Sns.Timestamp": "2017-07-05T20:01:21.366Z",
  "@iopipe/event-info.sns.Records[0].Sns.TopicArn": "arn:aws:sns:us-west-2:123456789000:ses_messages",
  "@iopipe/event-info.sns.Records[0].Sns.Type": "Notification",
  "@iopipe/event-info.sns.Records[0].Sns.UnsubscribeUrl": "https://sns.us-west-2.amazonaws.com/?Action=Unsubscribe&eifjccgihujihfhrchunfnglreichbrcljrnlvtbeked",
}
`;
